Adds CodeCache handler

ry opened the pull request:
Depends on #396

piscisaureus commented on src/deno_dir.rs:
```diff
@@ -0,0 +1,121 @@
+  // In the Go code this was called SrcDir. Renaming to Deps.
+  // Example: /Users/rld/.deno/deps/
+  Deps,
+  // In the Go code this was called CacheDir. Renaming to GEN.
```
Gen or GEN ?

ry replied:
Fixed.

piscisaureus commented on src/deno_dir.rs:
```diff
@@ -0,0 +1,121 @@
+}
+pub use self::Dirname::*;
+
+pub fn path(dirname: Dirname) -> &'static Path {
```
Using Path now, that's a step forward :)

piscisaureus commented on src/deno_dir.rs:
```diff
@@ -0,0 +1,121 @@
+  if unsafe { ROOT != None } {
+    return Ok(());
+  }
+  let home_dir = std::env::home_dir().expect("Could not get home directory.");
```
Would rather convert `home_dir` Path/PathBuf here and get rid of the unsafe block below.

ry replied:
It is a PathBuf here already https://doc.rust-lang.org/std/env/fn.home_dir.html

The unsafe is for the static global assignment.

piscisaureus replied:
Ah, I didn't know this was considered unsafe

piscisaureus commented on src/deno_dir.rs:
```diff
@@ -0,0 +1,121 @@
+
+// https://github.com/ry/deno/blob/golang/deno_dir.go#L32-L35
+pub fn cache_path(filename: &str, source_code: &str) -> PathBuf {
+  let cache_key = source_code_hash(filename, source_code);
```
So if a file's dependencies change, there's no need to recompile?
(Just asking, I don't know)

ry replied:
That's right. In the compiled code the requires will reference the original module specifier, not the specifier of the compiled file.

ry replied:
@piscisaureus Yeah, I don't think that happens... but let's see if we find any examples as we flesh out the tests.

piscisaureus commented on src/fs.rs:
```diff
@@ -0,0 +1,19 @@
+pub fn read_file_sync(path: &Path) -> std::io::Result<String> {
+  File::open(path).and_then(|mut f| {
+    let mut contents = String::new();
+    f.read_to_string(&mut contents)?;
```
This is not generally safe, because a file's contents are not necessarily valid utf-8.
We'll make it read to a buffer later? Otherwise consider renaming the function to `read_text_file_sync`.

ry replied:
Oh - I thought Strings were like `std::string`. Sigh. Will change...

piscisaureus replied:
This gonna stay as it is?

piscisaureus commented on src/handlers.rs:
```diff
@@ -21,9 +21,13 @@ fn test_url() {
-fn string_from_ptr(ptr: *const c_char) -> String {
-  let cstr = unsafe { CStr::from_ptr(ptr as *const i8) };
-  String::from(cstr.to_str().unwrap())
+// Help. Is there a way to do this without macros?
```
I don't think so -- an &str expects to store a reference to it's underlying String, but `*const c_char` isn't and can't be used as such (a String is not just a pointer, but also stores `length`, `capacity`, `refcount` etc.).
This is the reason you have to construct a new String here. 
If you make `str_from_ptr` a function then the String goes out of scope at the end of the function.
(you could also return both, but I suspect that might also defeat the purpose of this convenience wrapper).

ry commented on build_extra/flatbuffers/rust/rust_flatbuffer.gni:
```diff
@@ -0,0 +1,47 @@
+import("//build_extra/rust/rust.gni")
+
+# TODO(ry) "flatbuffer.gni" should be "flatbuffers.gni" we should be consistant
```
Done.

ry commented on src/deno_dir.rs:
```diff
@@ -0,0 +1,121 @@
+// https://github.com/ry/deno/blob/golang/deno_dir.go#L99-L111
+pub fn setup() -> std::io::Result<()> {
+  // Only setup once.
+  if unsafe { ROOT != None } {
```
Because it's a static global.

piscisaureus commented:
Mostly LGTM now. Strangely I didn manage to build this PR, but now that I'm trying again, I get this error (freshly checked out, cleaned, third_party reset and cleaned etc).

C:\Users\BertBelder\d\deno>python tools\build.py
C:\Users\BertBelder\d\deno\third_party\depot_tools\gn gen C:\Users\BertBelder\d\deno\out\debug
Traceback (most recent call last):
  File "C:/Users/BertBelder/d/deno/build/vs_toolchain.py", line 16, in <module>
    from gn_helpers import ToGNString
ImportError: No module named gn_helpers
ERROR at //build/config/win/visual_studio_version.gni:27:7: Script returned non-zero exit code.
      exec_script("../../vs_toolchain.py", [ "get_toolchain_dir" ], "scope")
      ^----------
Current dir: C:/Users/BertBelder/d/deno/out/debug/
Command: C:/Python27/python.exe -- C:/Users/BertBelder/d/deno/build/vs_toolchain.py get_toolchain_dir
Returned 1.
See //build/toolchain/win/BUILD.gn:8:1: whence it was imported.
import("//build/config/win/visual_studio_version.gni")
^----------------------------------------------------
See //BUILD.gn:9:1: which caused the file to be included.
group("all") {

piscisaureus commented:
Hmm, it seems that git creates the wrong kind of symlinks now that third_party has become a separate repo.

robbym commented:
Are the global mutables necessary? Can't all of this go into a struct that owns those globals and methods?

ry commented:
@robbym I agree it shouldn't be global. I've reworked it to be a standalone struct.

ry commented:
@piscisaureus PTAL. I've done some refactoring and added more tests.
1. DenoDir is now a standalone struct - instead of using global variables. This makes threaded testing possible and is also the right thing to do.
2. In order to store the DenoDir in the Deno struct and then retrieve it later, I've added `from_c` and `deno_get_data`. 
3.  Due to that `remove_dir_all`, you may find you don't have all the deps for windows. I didn't want to inadvertently add unused stuff.. so I'll leave it to you do add winapi, if it is indeed needed.

piscisaureus commented:
The test fails for me:

<details>
<summary>output</summary>
running 9 tests
test deno_dir::test_resolve_module ... ok
test deno_dir::test_cache_path ... ok
test deno_dir::test_source_code_hash ... ok
test deno_dir::test_code_cache ... ok
test deno_dir::test_src_file_to_url ... ok
test test_parse_core_args_1 ... ok
test test_parse_core_args_2 ... ok
test test_c_to_rust ... ok
test deno_dir::test_code_fetch ... FAILED

failures:

---- deno_dir::test_code_fetch stdout ----
        thread 'deno_dir::test_code_fetch' panicked at 'called `Result::unwrap()` on an `Err` value: ()', libcore\result.rs:945:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::windows::backtrace::unwind_backtrace
             at C:\projects\rust\src\libstd\sys\windows\backtrace\mod.rs:65
   1: std::sys_common::backtrace::print
             at C:\projects\rust\src\libstd\sys_common\backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at C:\projects\rust\src\libstd\panicking.rs:211
   3: std::panicking::default_hook
             at C:\projects\rust\src\libstd\panicking.rs:221
   4: std::panicking::rust_panic_with_hook
             at C:\projects\rust\src\libstd\panicking.rs:463
   5: std::panicking::begin_panic_fmt
             at C:\projects\rust\src\libstd\panicking.rs:350
   6: std::panicking::rust_begin_panic
             at C:\projects\rust\src\libstd\panicking.rs:328
   7: core::panicking::panic_fmt
             at C:\projects\rust\src\libcore\panicking.rs:71
   8: core::result::unwrap_failed<()>
             at C:\projects\rust\src\libcore\macros.rs:26
   9: core::result::Result<url::Url, ()>::unwrap<url::Url,()>
             at C:\projects\rust\src\libcore\result.rs:782
  10: test_rs_bin::deno_dir::DenoDir::resolve_module
             at .\src\deno_dir.rs:169
  11: test_rs_bin::deno_dir::DenoDir::code_fetch
             at .\src\deno_dir.rs:98
  12: test_rs_bin::deno_dir::test_code_fetch
             at .\src\deno_dir.rs:273
  13: test_rs_bin::__test::TESTS::{{closure}}
             at .\src\deno_dir.rs:264
  14: core::ops::function::FnOnce::call_once<closure,()>
             at C:\projects\rust\src\libcore\ops\function.rs:223
  15: alloc::boxed::{{impl}}::call_box<(),closure>
             at C:\projects\rust\src\liballoc\boxed.rs:638
  16: panic_unwind::__rust_maybe_catch_panic
             at C:\projects\rust\src\libpanic_unwind\lib.rs:105


failures:
    deno_dir::test_code_fetch

test result: FAILED. 8 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
</details>

ry commented:
@piscisaureus I believe I've fixed that using your add_root! macro.

piscisaureus commented on src/deno_dir.rs:
```diff
@@ -0,0 +1,398 @@
+    if cache_path.exists() {
+      Ok(())
+    } else {
+      let mut file = File::create(cache_path)?;
```
This is a race condition w.r.t to exists() -- probably should create the file in exclusive mode.
I also worry what might happen is there are two processes and one reads the cache file while the other is in the midst of writing it. 
Not suggesting you fix it as part of this PR, but it'll have to be done some time.

ry replied:
I added a todo comment.

piscisaureus commented on src/handlers.rs:
```diff
@@ -254,63 +67,67 @@ pub extern "C" fn handle_code_fetch(
+    let errmsg = format!("{}", err);
+    reply_error(d, cmd_id, &errmsg);
+  }
+  // null response indicates success.
```
Implicit null response is a foot gun. Probably better to explicitly say reply_null() if a handler succeeds without a return value.

ry replied:
Agree. I think this whole interface is going to change dramatically soon. This is just a first pass. I'll leave it as is now.

piscisaureus commented on src/main.rs:
```diff
@@ -89,14 +88,30 @@ fn set_flags(args: Vec<String>) -> Vec<String> {
+      ptr: 0 as *const binding::DenoC,
+      dir: deno_dir::DenoDir::new(None).unwrap(),
+    });
+    let deno: &'a mut Deno = Box::leak(deno_box);
```
Jeez... I can see how this makes for some sleepless nights.
Why is Deno made mutable here if it gets cast to *const one line below?
(Not saying it's wrong, I just don't understand this very well.)

A little comment on what's going on here would be in order too (as in -- mention that C-Deno needs a reference to Rust-Deno. And why this leak is necessary (is it really btw?).

ry replied:
It gets returned as mut (but my confidence that this is optimal code is low)
